rp2040-hal/src/spi.rs

```diff
@@ -78,53 +78,131 @@ impl<S: State, D: SpiDevice, const DS: u8> Spi<S, D, DS> {
-            // So clamping a u32::MAX is fine here...
-            if freq_in < ((prescale_option + 2) * 256).saturating_mul(baudrate) {
-                prescale = prescale_option as u8;
-                break;
-            }
-        }
+
+        let settings = calc_spi_clock_divider_settings_for_baudrate(freq_in, baudrate);
 
         // We might not find a prescale value that lowers the clock freq enough, so we leave it at max
-        debug_assert_ne!(prescale, u8::MAX);
-
-        // Find largest post-divide which makes output <= baudrate. Post-divide is
-        // an integer in the range 0 to 255 inclusive.
-        for postdiv_option in (1..=255u8).rev() {
-            if freq_in / (prescale as u32 * postdiv_option as u32) > baudrate {
-                postdiv = postdiv_option;
-                break;
-            }
-        }
+        debug_assert_ne!(settings.prescale, u8::MAX);
+
+        self.set_baudrate_from_settings(&settings);
+
+        // Return the frequency we were able to achieve
+        use fugit::RateExtU32;
+        (freq_in / (settings.prescale as u32 * (1 + settings.postdiv as u32))).Hz()
+    }
 
+    /// Set the baudrate using a previously calculated [SpiClockDividerSettings]
+    pub fn set_baudrate_from_settings(&mut self, settings: &SpiClockDividerSettings) {
         self.device
             .sspcpsr
-            .write(|w| unsafe { w.cpsdvsr().bits(prescale) });
+            .write(|w| unsafe { w.cpsdvsr().bits(settings.prescale) });
         self.device
             .sspcr0
-            .modify(|_, w| unsafe { w.scr().bits(postdiv) });
+            .modify(|_, w| unsafe { w.scr().bits(settings.postdiv) });
+    }
 
-        // Return the frequency we were able to achieve
-        use fugit::RateExtU32;
-        (freq_in / (prescale as u32 * (1 + postdiv as u32))).Hz()
+    /// Set the mode
+    ///
+    /// Momentarily disables / enables the device so be careful of truncating ongoing transfers.
+    pub fn set_mode(&mut self, mode: Mode) {
+        // disable the device
+        self.device.sspcr1.modify(|_, w| w.sse().clear_bit());
+
+        // Set the polarity and phase
+        self.device.sspcr0.modify(|_, w| {
+            w.spo()
+                .bit(mode.polarity == Polarity::IdleHigh)
+                .sph()
+                .bit(mode.phase == Phase::CaptureOnSecondTransition)
+        });
+
+        // enable the device
+        self.device.sspcr1.modify(|_, w| w.sse().set_bit());
+    }
+
+    /// Wait until all operations have completed and the bus is idle.
+    pub fn flush(&self) -> Result<(), nb::Error<Infallible>> {
+        if self.device.sspsr.read().bsy().bit() {
+            Err(nb::Error::WouldBlock)
+        } else {
+            Ok(())
+        }
     }
 }
 
+/// Clock divider settings
+pub struct SpiClockDividerSettings {
+    /// The prescaler for writing to sspcpsr
+    pub prescale: u8,
+    /// The postdiv for writing to sspcr0
+    pub postdiv: u8,
+}
```

Comment:
Making fields public is a strong commitment (see rust's [future proofing](https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private) guidelines).

Also, I'm not even sure allowing users to set them to arbitrary values is a good idea since they'd need to be validated before set in the registers (eg only even values are allowed in `CPSDVSR`).